comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -0,0 +1,184 @@
+        Self { connection, capacity }
+    }
+
+    /// Inserts and returns Ok(true) if the item already existed and Ok(false) if it didn't
+    pub async fn insert_body_hash_if_unique(&self, body_hash: Vec<u8>) -> Result<bool, StorageError> {
+        let body_hash_string = hex::to_hex(&body_hash.as_bytes());
+        match self.insert_body_hash(body_hash_string.clone()).await {
+            Ok(val) => {
+                if val == 0 {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Unable to insert new entry into message dedup cache"
+                    );
+                }
+                Ok(false)
+            },
+            Err(e) => match e {
+                StorageError::UniqueViolation(_) => match self.update_number_of_hits(body_hash_string).await {
+                    Ok(_) => Ok(true),
+                    Err(e) => Err(e),
+                },
+                _ => Err(e),
+            },
+        }
+    }
+
+    /// Trims the dedup cache to the configured limit by removing the oldest entries
+    pub async fn truncate(&self) -> Result<usize, StorageError> {
+        let capacity = self.capacity;
+        self.connection
+            .with_connection_async(move |conn| {
+                let mut num_removed = 0;
+                let msg_count = dedup_cache::table
+                    .select(dsl::count(dedup_cache::id))
+                    .first::<i64>(conn)? as usize;
+                // Hysteresis added to minimize database impact
+                if msg_count > capacity {
+                    let remove_count = msg_count - capacity;
+                    let message_ids: Vec<i32> = dedup_cache::table
+                        .select(dedup_cache::id)
+                        .order_by(dedup_cache::stored_at.asc())
+                        .limit(remove_count as i64)
+                        .get_results(conn)?;
+                    num_removed = diesel::delete(dedup_cache::table)
+                        .filter(dedup_cache::id.eq_any(message_ids))
+                        .execute(conn)?;
+                }
+                debug!(
+                    target: LOG_TARGET,
+                    "Message dedup cache: msg_count {}, capacity {}, num_removed {}", msg_count, capacity, num_removed,
+                );
+                Ok(num_removed)
+            })
+            .await
+    }
+
+    async fn insert_body_hash(&self, body_hash: String) -> Result<usize, StorageError> {
+        self.connection
+            .with_connection_async(move |conn| {
+                let insert_result = diesel::insert_into(dedup_cache::table)
+                    .values(UpdateDedupCacheSql {
+                        body_hash: Some(body_hash.clone()),
+                        number_of_hits: Some(1),
+                        last_hit_at: Some(DedupCacheDatabase::formatted_naive_date_time()?),
+                    })
+                    .execute(conn)
+                    .map_err(Into::into);
```

Comment:
```suggestion
                    ?;
```

Reply:
This converts `QueryResult<usize>` into `Result<usize, StorageError>`; the suggested change results in `usize`.
The `UniqueViolation` error is used as a feature to determine if the entry is already present.

Reply:
I think the `?` operator is the same as `map_err(Into::into)`.

Reply:
It  results in `usize` instead of `Result<usize, StorageError>`

Reply:
You can just remove the `map_err(Into::into)`. The conversion is redundant here because it is already a `diesel::result::Error`, which is the type the match expects.  `Err(e) => Err(e.into()),` converts the error into a `StorageError` so that it can be returned.